persistent/Database/Persist/Sql/TH.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE TemplateHaskellQuotes #-}
+
+module Database.Persist.Sql.TH
+  ( fromN
+  , generateRawSqlHelper
+  , generateRawSqlInstances
+  , toN
+  )
+where
+
+import Control.Monad
+import Language.Haskell.TH
+
+#if MIN_VERSION_template_haskell(2,16,0)
+prepTupE :: a -> Maybe a
+prepTupE = Just
+#else
+prepTupE :: a -> a
+prepTupE = id
+#endif
+
+minTupleSize :: Int
+minTupleSize = 3 -- instance for 2 is already written
+
+maxTupleSize :: Int
+maxTupleSize = 62 -- GHC limitation
+
+pairedList :: [a] -> [[a]]
+pairedList [] = []
+pairedList [x] = [[x]]
+pairedList (x:y:rest) = [x, y] : pairedList rest
+
+flatTupleT :: [Name] -> Type
+flatTupleT names = foldl1 AppT $ (TupleT $ length names) : map VarT names
+
+nestedTupleT :: [Name] -> Type
+nestedTupleT names =
+   foldl1 AppT $ (TupleT $ (length names + 1) `div` 2) : map tupleOrSingleT (pairedList names)
+
+tupleOrSingleT :: [Name] -> Type
+tupleOrSingleT [a, b] = AppT (AppT (TupleT 2) (VarT a)) (VarT b)
+tupleOrSingleT [a] = VarT a
+tupleOrSingleT x = error $ "tupleOrSingleT failed to convert: " <> show x
+
+fromN :: Int -> Q (Type, Exp)
+fromN n = do
+    names <- replicateM n $ newName "x"
+    let arg = [TupP . map VarP $ names]
+    pure ( AppT (AppT ArrowT $ flatTupleT names) (nestedTupleT names)
+         , LamE arg . TupE . map (prepTupE . tupleOrSingleE) $ pairedList names
+         )
+
+    where
+        tupleOrSingleE :: [Name] -> Exp
+        tupleOrSingleE [a, b] = TupE $ map (prepTupE . VarE) [a, b]
+        tupleOrSingleE [a] = VarE a
+        tupleOrSingleE x = error $ "tupleOrSingleE failed to convert: " <> show x
+
+toN :: Int -> Q (Type, Exp)
+toN n = do
+    names <- replicateM n $ newName "x"
+    let arg = [TupP . map tupleOrSingleP $ pairedList names]
+    pure ( AppT (AppT ArrowT $ nestedTupleT names) (flatTupleT names)
+         , LamE arg . TupE $ map (prepTupE . VarE) names
+         )
+
+    where
+        tupleOrSingleP :: [Name] -> Pat
+        tupleOrSingleP [a, b] = TupP $ map VarP [a, b]
+        tupleOrSingleP [a] = VarP a
+        tupleOrSingleP x = error $ "tupleOrSingleP failed to convert: " <> show x
+
+rawSqlInstanceN :: Int -> Q Dec
+rawSqlInstanceN n = do
+    names <- replicateM n $ newName "x"
+
+    let constraints = map (\name -> AppT (ConT $ mkName "RawSql") (VarT name)) names
+        tuple = AppT (ConT $ mkName "RawSql") . foldl1 AppT $ (TupleT $ length names) : map VarT names
+
+    pure $ InstanceD Nothing constraints tuple
+        [ FunD (mkName "rawSqlCols") [Clause [] (NormalB rawSqlColsBody) []]
+        , FunD (mkName "rawSqlColCountReason") [Clause [] (NormalB rawSqlColCountReasonBody) []]
+        , FunD (mkName "rawSqlProcessRow") [Clause [] (NormalB rawSqlProcessRowBody) []]
```

Comment:
```suggestion
        [ FunD ('rawSqlCols) [Clause [] (NormalB rawSqlColsBody) []]
        , FunD ('rawSqlColCountReason) [Clause [] (NormalB rawSqlColCountReasonBody) []]
        , FunD ('rawSqlProcessRow) [Clause [] (NormalB rawSqlProcessRowBody) []]
```

Reply:
So, on another note, you can *also* make this a QuasiQuote.

```haskell
[d|
    instance $(pure constraints) => $(pure tuple) where
        rawSqlCols = $(rawSqlColsBody)
        rawSqlColCountReason = $(rawSqlColCountReasonBody)
        rawSqlProcessRow = $(rawSqlProcessRowBody)
```

The expression in a splice needs to be `Q Exp` (or something that can be automatically `lift :: Lift a => a -> Q Exp`. If you use the QQ for the `rawSqlColsBody`, then it returns a `Q Exp`, and you can use it directly.

Or, you could just have the instance completely inline with the QQ, too.

```haskell
[d|
    instance $(pure constraints) => $(pure tuple) where
        rawSqlCols = \e -> rawSqlCols e . $(varE $ mkName $ "from" ++ show n)
        -- etc...
  |]
```

Reply:
When I try to quote the instance declaration like this
```haskell
    [d|
      instance $(pure constraints) => $(pure tuple) where
          rawSqlCols = \e -> rawSqlCols e . $(fromOrToN From n)
          rawSqlColCountReason = rawSqlColCountReason . $(fromOrToN From n)
          rawSqlProcessRow = fmap $(fromOrToN To n) . rawSqlProcessRow
      |]
```
I get this back from -ddump-splices, where the names of functions in the instance look like they're not matching up with the names within the expression (`rawSqlCols_02y` /= `rawSqlCols`). This then leads to a bunch of errors in compiling what TH generated (redefining functions, missing names, etc.)
```haskell
instance (RawSql x_a48o6, RawSql x_a48o7, RawSql x_a48o8) =>
             RawSql (x_a48o6, x_a48o7, x_a48o8) where
      rawSqlCols_02y = \ e_a48o9 -> (rawSqlCols e_a48o9 . from3)
      rawSqlColCountReason_02y = (rawSqlColCountReason . from3)
      rawSqlProcessRow_02y = (fmap to3 . rawSqlProcessRow)
```
However, trying to fix this issue by putting a `Name` there directly (trying various ways)
```haskell
    [d|
      instance $(pure constraints) => $(pure tuple) where
          $(pure 'rawSqlCols) = \e -> rawSqlCols e . $(fromOrToN From n)
          $(pure $ mkName "rawSqlColCountReason") = rawSqlColCountReason . $(fromOrToN From n)
          $(lookupValueName "rawSqlProcessRow") = fmap $(fromOrToN To n) . rawSqlProcessRow
      |]
```
yields errors like
```haskell
Pattern bindings (except simple variables) not allowed in instance declaration:
        $(pure 'rawSqlCols) = \ e -> rawSqlCols e . $(fromOrToN From n)
```

I was able to get things working by handwriting TH `FunD` constructors over quoted expressions, but not when quoting the whole instance, so I've probably misunderstood how that works